lightning/src/util/ser_macros.rs

```diff
@@ -8,81 +8,153 @@
 // licenses.
 
 macro_rules! encode_tlv {
-	($stream: expr, {$(($type: expr, $field: expr)),*}, {$(($optional_type: expr, $optional_field: expr)),*}) => { {
+	($stream: expr, $type: expr, $field: expr, required) => {
+		BigSize($type).write($stream)?;
+		BigSize($field.serialized_length() as u64).write($stream)?;
+		$field.write($stream)?;
+	};
+	($stream: expr, $type: expr, $field: expr, vec_type) => {
+		encode_tlv!($stream, $type, ::util::ser::VecWriteWrapper(&$field), required);
+	};
+	($stream: expr, $optional_type: expr, $optional_field: expr, option) => {
+		if let Some(ref field) = $optional_field {
+			BigSize($optional_type).write($stream)?;
+			BigSize(field.serialized_length() as u64).write($stream)?;
+			field.write($stream)?;
+		}
+	};
+}
+
+macro_rules! encode_tlv_stream {
+	($stream: expr, {$(($type: expr, $field: expr, $fieldty: ident)),*}) => { {
 		#[allow(unused_imports)]
-		use util::ser::BigSize;
-		// Fields must be serialized in order, so we have to potentially switch between optional
-		// fields and normal fields while serializing. Thus, we end up having to loop over the type
-		// counts.
-		// Sadly, while LLVM does appear smart enough to make `max_field` a constant, it appears to
-		// refuse to unroll the loop. If we have enough entries that this is slow we can revisit
-		// this design in the future.
-		#[allow(unused_mut)]
-		let mut max_field: u64 = 0;
-		$(
-			if $type >= max_field { max_field = $type + 1; }
-		)*
-		$(
-			if $optional_type >= max_field { max_field = $optional_type + 1; }
-		)*
-		#[allow(unused_variables)]
-		for i in 0..max_field {
-			$(
-				if i == $type {
-					BigSize($type).write($stream)?;
-					BigSize($field.serialized_length() as u64).write($stream)?;
-					$field.write($stream)?;
-				}
-			)*
+		use {
+			ln::msgs::DecodeError,
+			util::ser,
+			util::ser::BigSize,
+		};
+
+		#[allow(unused_mut, unused_variables, unused_assignments)]
+		{
+			let mut last_seen: Option<u64> = None;
 			$(
-				if i == $optional_type {
-					if let Some(ref field) = $optional_field {
-						BigSize($optional_type).write($stream)?;
-						BigSize(field.serialized_length() as u64).write($stream)?;
-						field.write($stream)?;
-					}
+				encode_tlv!($stream, $type, $field, $fieldty);
```

Comment:
May be worth pulling this out into a separate repetition and making this repetition `#[cfg(debug_assertions)]`-enabled in its own scope. Release builds might be better optimized since `last_seen` wouldn't exist. I think the `allow` wouldn't be needed either.

Reply:
Is this the implementation you were thinking of? https://i.imgur.com/G3g1BNz.png Seems it'll still need the `allow`

Reply:
Eh, if there's one thing I have confidence in LLVM for, it's optimizing away purely-on-stack variables which are written too and never read (and branches with no effect).

Reply:
I was thinking `last_seen` would go inside the second `#[cfg(debug_assertions)]` scope, leaving only one `#[cfg(debug_assertions)]`. But maybe it's still problematic?

Reply:
> I was thinking `last_seen` would go inside the second `#[cfg(debug_assertions)]` scope, leaving only one `#[cfg(debug_assertions)]`. But maybe it's still problematic?

Ah, for some reason that was less obviously correct to me, but it does seem to be right according to `-Z trace-macros`. Still needs the `allow` but otherwise done